runtime/ast/walk.go

```diff
@@ -0,0 +1,62 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2021 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ast
+
+type Walker interface {
+	Walk(element Element) Walker
+}
+
+// Walk traverses an AST in depth-first order:
+// It starts by calling walker.Walk(element);
+// If the returned walker is nil,
+// child elements are not walked.
+// If the returned walker is not-nil,
+// then Walk is invoked recursively on this returned walker
+// for each of the non-nil children of the element,
+// followed by a call of Walk(nil) on the returned walker.
+//
+func Walk(walker Walker, element Element) {
+	if walker = walker.Walk(element); walker == nil {
```

Comment:
Could the initial `walker` parameter be nil?

Reply:
Good question! I'll document that it can't, only the walker returned from a walker

Reply:
Added in 7aaf2da8e8e458c415d718651ddd88dda30871fd